misc/python/materialize/output_consistency/validation/validation_message.py

```diff
@@ -72,33 +73,34 @@ def __str__(self) -> str:
 class ValidationError(ValidationMessage):
     def __init__(
         self,
+        query_template: QueryTemplate,
         error_type: ValidationErrorType,
         message: str,
+        strategy1: EvaluationStrategy,
+        strategy2: EvaluationStrategy,
         description: Optional[str] = None,
         value1: Optional[str] = None,
         value2: Optional[str] = None,
-        strategy1: Optional[EvaluationStrategy] = None,
-        strategy2: Optional[EvaluationStrategy] = None,
         sql1: Optional[str] = None,
         sql2: Optional[str] = None,
+        col_index: Optional[int] = None,
         location: Optional[str] = None,
     ):
         super().__init__(message, description)
+        self.query_template = query_template
         self.error_type = error_type
         self.value1 = value1
         self.value2 = value2
         self.strategy1 = strategy1
         self.strategy2 = strategy2
         self.sql1 = sql1
         self.sql2 = sql2
+        self.col_index = col_index
         self.location = location
 
         if value1 is None and value2 is not None:
```

Comment:
These assertions on the inputs smell like there should be two different methods to initialize it instead, so that the constraint is entirely expressed through the parameter types.

Reply:
I had a look at it and figured out that the "value" and "sql" properties are actually always set. Therefore, I changed the type from Optional and simplified the code. Thanks for the hint!

✅